feat(identity): add backend pagination to all List views

marstamm opened the pull request:
## Description
This PR adds backend pagination to Identity pages.

 - Requests are limited by page size
 - Sorting is configurable and also done on the backend. Uses local loading state in the `EntityList`.
 - Search is done on the backend. Introduces local "No data" state to the `EntityList`.
   As decided in https://github.com/camunda/camunda/issues/33972#issuecomment-3073637976 , Member views will not support searching

## Related issues

closes https://github.com/camunda/camunda/issues/33972

nithinssabu reviewed (commented):
Great work overall, @marstamm 👏 

This has been a bit of a challenging review for me, as I'm still getting familiar with the codebase - and the size of the PR made it tougher to digest in one go. I have completed a partial review for now and will continue the rest tomorrow. 

Most of my comments are marked as take it or leave it (🔧), but in a codebase I'm responsible for, several of them would be must-fix items (❌). Please take a closer look and assess what fits best in this context.

For future PRs, consider breaking down larger implementations into smaller, focussed changes. This makes the review process more manageable, leads to more thoughtful feedback, and ultimately results in higher-quality code.

marstamm commented on identity/client/src/components/entityList/EntityList.tsx:
```diff
@@ -127,16 +132,39 @@ const EntityList = <D extends EntityData>({
-  const [page, setPage] = useState(1);
-  const [pageSize, setPageSize] = useState(PAGINATION_MAX_PAGE_SIZE);
+
+  // Search handling
+  const [search, setSearchState] = useState<string>();
+
+  useEffect(() => {
+    if (!searchKey) {
+      return;
+    }
+
+    if (typeof search !== "string") {
+      return;
+    }
+
+    if (search.length === 0) {
+      debounce(() => setSearch(undefined));
+      return;
+    }
+
+    debounce(() => setSearch({ [searchKey]: search }));
+  }, [debounce, search, searchKey]);
```
It does make sense to pull search behavior into a separate Component. With that, I believe the complexity of the (Search) Component will be small enough, so we don't need a custom hook to handle the search, unless we find another Place that needs the same behavior.

nithinssabu replied:
> the complexity of the (Search) Component will be small enough, so we don't need a custom hook to handle the search

Agree 👍

marstamm replied:
Done with [96caff0](https://github.com/camunda/camunda/pull/35887/commits/96caff061c65a38d551cb4b8ac237b4bb04a5f1f)

marstamm commented on identity/client/src/components/entityList/EntityList.tsx:
```diff
@@ -419,30 +455,31 @@ const EntityList = <D extends EntityData>({
-          )}
+            )}
+          </>
+          {!!pageData?.totalItems &&
```
We need the cast to boolean here. `totalItems` is a `number`. If we want to hide the section completely, we need a boolean, as React accepts `0` as a valid child and will render it as Text in the DOM tree

nithinssabu replied:
Oh, right 🙂 That makes sense - thanks for the explanation!

marstamm commented on identity/client/src/components/entityList/EntityList.tsx:
```diff
@@ -127,16 +132,39 @@ const EntityList = <D extends EntityData>({
+
+    if (typeof search !== "string") {
+      return;
+    }
```
I could be undefined, but we can handle this better. Adjusted in [96caff0](https://github.com/camunda/camunda/pull/35887/commits/96caff061c65a38d551cb4b8ac237b4bb04a5f1f)

nithinssabu commented on identity/client/src/utility/api/hooks/usePaginatedApi.ts:
```diff
@@ -0,0 +1,58 @@
+
+export default usePaginatedApi;
+
+const mergeParams = (
+  custom = {} as Record<string, unknown>,
+  page = {} as PaginationRequestParams,
+): Record<string, unknown> => {
+  const result = { ...custom } as Record<string, unknown>;
+
+  for (const key of Object.keys(page)) {
+    const typedKey = key as keyof PaginationRequestParams;
+
+    if (result[typedKey] !== undefined) {
+      result[typedKey] = {
+        ...result[typedKey],
+        ...page[typedKey],
+      };
+    } else {
+      result[typedKey] = page[typedKey];
+    }
+  }
+
+  return result;
+};
```
❓ Why is a deep merge used here, but a shallow merge is used in `usePaginatedApiCall`?

nithinssabu commented on identity/client/src/utility/api/groups/index.ts:
```diff
@@ -29,12 +30,15 @@ type SearchGroupsParams = {
-> = (filterParams) => {
-  const params = filterParams?.groupIds
-    ? { filter: { groupId: { $in: filterParams.groupIds } } }
+  Partial<SearchGroupsParams & PageSearchParams> | undefined
```
Yes, you're right - thanks for pointing that out! I'm entirely new to TS as well.

nithinssabu commented on identity/client/src/components/entityList/SearchBar.tsx:
```diff
@@ -0,0 +1,63 @@
+  debounce = 300,
+}: SearchBarProps): ReturnType<FC> {
+  // Search handling
+  const [search, setSearchState] = useState<string>();
```
❌ Since `search` is initially `undefined`, the following warning appears in the console:
```
hook.js:608 Warning: A component is changing an uncontrolled input to be controlled. This is likely caused by the value changing from undefined to a defined value, which should not happen. Decide between using a controlled or uncontrolled input element for the lifetime of the component. 
```

To avoid this, use a default value of empty string ("") so the input remains controlled from the start.

nithinssabu commented on identity/client/src/pages/authorizations/AuthorizationsList.tsx:
```diff
@@ -17,20 +16,21 @@ import EntityList from "src/components/entityList";
   data: SearchResponse<Authorization> | null;
   loading: boolean;
   reload: () => unknown;
+  paginationProps: Partial<Omit<ReturnType<typeof usePaginatedApi>, "data">>;
```
That was my guess as well - I was mainly confused that only "data" was being omitted and it was probably working because of the "Partial" wrapping. Now it looks cleaner!

That said, I'm not convinced if these type gymnastics are really worth the effort. It's just slowing us down without providing much value. This is off-topic, but I would love to hear your take on this after working with Typescript. We had a heated debate around it on the Web Modeler project a few months ago and decided against using Typescript - and I'm feeling even more and more confident in that decision.

nithinssabu commented on identity/client/src/components/entityList/SearchBar.tsx:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    debounceFn(() => onSearch({ [searchKey]: search }));
+  }, [debounce, search, searchKey]);
```
❌ The dependency array should be
```suggestion
  }, [debounceFn, onSearch, search, searchKey]);
```

vitorwtavares commented on identity/client/src/components/entityList/SearchBar.tsx:
```diff
@@ -0,0 +1,63 @@
+  onSearch,
+  debounce = 300,
+}: SearchBarProps): ReturnType<FC> {
+  // Search handling
```
We can remove this comment.

vitorwtavares reviewed (approved):
@nithinssabu had already pointed out a few minor things which I agree with, other than that there's a comment we can remove.

Overall LGTM, good job!

nithinssabu commented on identity/client/src/utility/api/hooks/usePaginatedApiCall.ts:
```diff
@@ -0,0 +1,46 @@
+    [_call, pageParams],
+  );
+
+  useEffect(() => {
+    if (result.data?.page) {
+      setPageResult(result.data.page);
+    }
+  }, [result.data]);
```
❓ Could you clarify why a `useEffect` is used here to handle `result.data.page`, while `usePaginatedApi` manages similar logic without a `useEffect`? Is the difference intentional or could both hooks be unified for consistency?

marstamm replied:
Great catch! You are right, it is not needed here. We don't even need the state, we can just map data types directly in the return. This makes this hook much simpler

I thought state was needed since we handle multiple API calls instead of a single one, but that is already covered by `useApiCall`, no need to duplicate it.

nithinssabu reviewed (approved):
🔧 In my macbook, I need to scroll to access the pagination controls. This can be avoided by reducing the default page size to 10 and would improve usability by keeping the controls visible without scrolling.

https://github.com/user-attachments/assets/f0230b9d-1c7a-4fb7-9a0d-21505396aab2

marstamm commented on identity/client/src/utility/api/hooks/usePagination.ts:
```diff
@@ -0,0 +1,151 @@
+
+export type PageResult = {
+  totalItems: number;
+  startCursor: string;
+  endCursor: string;
+  hasMoreTotalItems: boolean;
```
If we consider removing them from the Backend, let's remove them here as well

marstamm commented:
Thank you both for the thorough reviews! I have addressed your comments and will squash the commits so we have a clean commit history on main.

>CI is also breaking, which should be fixed 

I think the e2e tests are just flaky on CI. They work fine locally:

<img width="670" height="186" alt="image" src="https://github.com/user-attachments/assets/fec82024-db02-484e-8c48-c2aee08e3de3" />

I tried increasing the timeout in [d9f2b54](https://github.com/camunda/camunda/pull/35887/commits/d9f2b54b3255167b300a2f967c185ba4736d1d2a), but I'm not sure if this will actually fix it. That's usually the problem with e2e tests.

If increasing the timeout does not fix it, how do you want to proceed @vitorwtavares ?

> and a rebase is probably needed here

Done.

marstamm commented:
>If increasing the timeout does not fix it, how do you want to proceed

I gave up trying to fix it the flaky tests and decided to just skip them. Let me know if you want to take a shot at it @vitorwtavares

vitorwtavares commented:
> > If increasing the timeout does not fix it, how do you want to proceed
> 
> I gave up trying to fix the flaky tests and decided to just skip them. Let me know if you want to take a shot at it @vitorwtavares

@marstamm The part regarding Tenants is because Tenants now are hidden by default by a new env variable, so that needs to be fixed, very similar to the OIDC flag to be able to unskip Mapping Rules tests. [I created a issue to fix that](https://github.com/camunda/camunda/issues/36092). Weird part is that I created the PR that added this new env var and the E2E tests didn't fail on my PR, but I'm pretty sure it's related to that.

Regarding authorizations, the tests were failing due to flakyness on the `waitForItem` utility, I found the issue and already implemented a solution, I'll open a PR tomorrow and assign you as a reviewer.
